src/main/java/build/buildfarm/worker/UploadManifest.java

```diff
@@ -0,0 +1,338 @@
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/** UploadManifest adds output metadata to a {@link ActionResult}. */
+/** FIXME move into worker implementation and implement 'fast add' with this for sharding */
+class UploadManifest {
+  private static final String ERR_NO_SUCH_FILE_OR_DIR = " (No such file or directory)";
+
+  private static final LinkOption[] NO_LINK_OPTION = new LinkOption[0];
+  // This isn't generally safe; we rely on the file system APIs not modifying the array.
+  private static final LinkOption[] NOFOLLOW_LINKS_OPTION =
+      new LinkOption[] { LinkOption.NOFOLLOW_LINKS };
+
+  private final DigestUtil digestUtil;
+  private final ActionResult.Builder result;
+  private final Path execRoot;
+  private final boolean allowSymlinks;
+  private final int inlineContentLimit;
+  private final Map<Digest, Path> digestToFile;
+  private final Map<Digest, Chunker> digestToChunkers;
+  private transient int inlineContentBytes;
+
+  /**
+   * Create an UploadManifest from an ActionResult builder and an exec root. The ActionResult
+   * builder is populated through a call to {@link #addFile(Digest, Path)}.
+   */
+  public UploadManifest(
+      DigestUtil digestUtil,
+      ActionResult.Builder result,
+      Path execRoot,
+      boolean allowSymlinks,
+      int inlineContentLimit) {
+    this.digestUtil = digestUtil;
+    this.result = result;
+    this.execRoot = execRoot;
+    this.allowSymlinks = allowSymlinks;
+    this.inlineContentLimit = inlineContentLimit;
+
+    this.digestToFile = new HashMap<>();
+    this.digestToChunkers = new HashMap<>();
+    this.inlineContentBytes = 0;
+  }
+
+  /**
+   * Add a collection of files or directories to the UploadManifest. Adding a directory has the
+   * effect of 1) uploading a {@link Tree} protobuf message from which the whole structure of the
+   * directory, including the descendants, can be reconstructed and 2) uploading all the
+   * non-directory descendant files.
+   *
+   * <p>Attempting to a upload symlink results in a {@link
+   * IllegalStateException}, since cachable actions shouldn't emit symlinks.
+   */
+  public void addFiles(Iterable<Path> files, CASInsertionPolicy policy)
+      throws IllegalStateException, IOException, InterruptedException {
+    for (Path file : files) {
+      // TODO(ulfjack): Maybe pass in a SpawnResult here, add a list of output files to that, and
+      // rely on the local spawn runner to stat the files, instead of statting here.
+      FileStatus stat = statIfFound(file, /* followSymlinks= */ false);
```

Comment:
Instead you could split the addFile and addDirectory methods which would allow you to make the expected output type check that we discussed.